Update existing workspaces

Molyna opened the pull request:
This should now update as needed on workspace imported through swagger.

It also deletes components that aren't in the swagger file and adds ones missing while keeping all external references.

kristianhelgesen commented:
Fikk denne da jeg prøvde å overskrive petstore noen ganger

Client configuration:  {:url http://localhost:8080, :options {:headers {Authorization Token token=08b5de5fa8b042b6a726c2cf24dbc2b1, Content-Type application/json, User-Agent ardoq-clojure-client}, :timeout 2000, :query-params {:org piedpiper}}}
{:url http://localhost:8080, :options {:headers {Authorization Token token=08b5de5fa8b042b6a726c2cf24dbc2b1, Content-Type application/json, User-Agent ardoq-clojure-client}, :timeout 2000, :query-params {:org piedpiper}}}
clojure.lang.ArityException: Wrong number of args (2) passed to: common/save-models
	at clojure.lang.AFn.throwArity(AFn.java:429)
	at clojure.lang.AFn.invoke(AFn.java:36)
	at ardoq.swagger.update_swagger2$create_component.invoke(update_swagger2.clj:12)
	at ardoq.swagger.update_swagger2$update_components$fn__10993.invoke(update_swagger2.clj:29)
	at clojure.core.protocols$fn__6074.invoke(protocols.clj:79)
	at clojure.core.protocols$fn__6031$G__6026__6044.invoke(protocols.clj:13)
	at clojure.core$reduce.invoke(core.clj:6289)
	at ardoq.swagger.update_swagger2$update_components.invoke(update_swagger2.clj:25)
	at ardoq.swagger.update_swagger2$update_workspace.invoke(update_swagger2.clj:146)
	at ardoq.swagger.swagger_v2$get_info.invoke(swagger_v2.clj:142)
	at ardoq.swagger.swagger_v2$import_swagger2.invoke(swagger_v2.clj:160)
	at ardoq.swagger.api$version2.invoke(api.clj:55)
	at ardoq.swagger.api$get_spec.invoke(api.clj:68)
	at ardoq.swagger.api$swagger_api$fn__11268.invoke(api.clj:89)
	at compojure.core$make_route$fn__7526.invoke(core.clj:104)
	at compojure.core$wrap_route_middleware$fn__7522.invoke(core.clj:94)
	at compojure.core$if_route$fn__7504.invoke(core.clj:41)
	at compojure.core$if_method$fn__7496.invoke(core.clj:27)
	at compojure.core$routing$fn__7532.invoke(core.clj:118)
	at clojure.core$some.invoke(core.clj:2515)
	at compojure.core$routing.doInvoke(core.clj:118)
	at clojure.lang.RestFn.applyTo(RestFn.java:139)
	at clojure.core$apply.invoke(core.clj:626)
	at compojure.core$routes$fn__7536.invoke(core.clj:123)
	at ring.middleware.session$wrap_session$fn__9208.invoke(session.clj:98)
	at ring.middleware.params$wrap_params$fn__8453.invoke(params.clj:64)
	at org.httpkit.server.HttpHandler.run(RingHandler.java:91)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

Molyna commented:
Fixed arity issue and then the fixed a deletion issue in regards to references in swagger 2.

Updated error response when the swagger url responds with 404 (used to just give a "Unkonwn exception message"

Should be fully set now but needs code-review from either @expez or @ebaxt

expez commented on src/ardoq/swagger/api.clj:
```diff
@@ -100,9 +100,14 @@
-             {:status 406
-              :headers {"Content-Type" "application/json"}
-              :body (json/write-str {:error (-> e ex-data :causes)})})
+             (println (= 404 (-> e ex-data :status)))
```
leftover println

expez commented on src/ardoq/swagger/client.clj:
```diff
@@ -81,8 +81,23 @@
+(defn find-aggregated [resource client]
+  (let [url (str (:url client) "/api/" (resource-path resource) "/" (:_id resource) "/aggregated")
+        {:keys [status body]} (http/get url (:options client))]
+    (cond
```
cond -> if

expez commented on src/ardoq/swagger/client.clj:
```diff
@@ -81,8 +81,23 @@
+  (let [url (str (:url client) "/api/" (resource-path resource) "/")
+        {:keys [status body]} (http/get url (:options client))]
+    (cond
+     (ok? status) (filter #(= (:rootWorkspace %) root-id) 
```
cond -> if

Molyna commented:
This pull request will now fix #8 and #9

Molyna commented:
Putting this on hold to create automated tests.

Molyna commented:
This can be taken tested again now as tests are complete. Need to setup variables for CircleCI so the tests can run. If you can look at it @kristianhelgesen @ebaxt @expez it would be great and we could push this into prod. 

We've uploaded to the repo 150 different swagger files now. Big question if we wanna delete some of them and just keep 5-20 that we wanna use on the tests (5 is probably more than enough as long as they are varied enough).

I am gonna run all 150 files in a test overnight today and see if we get any faults anywhere. (There are about 200 000 lines in the files in total so will take a few minutes to get them all in.

Molyna commented:
Ran the tests, 5 files failed. Only figured out which file it was in one case but was due to lazyness of clojure.

Will run again tonight to be sure latest fix fixes all of them. Otherwise we are ready to go.

Molyna commented:
All files and tests on files run now. This should be ready.
